spacy/morphology.pxd

```diff
@@ -1,21 +1,40 @@
-from cymem.cymem cimport Pool
-from preshed.maps cimport PreshMap
 cimport numpy as np
-from libc.stdint cimport uint64_t
+from libc.stdint cimport uint32_t, uint64_t
+from libcpp.unordered_map cimport unordered_map
+from libcpp.vector cimport vector
+from libcpp.memory cimport unique_ptr
+from libcpp.utility cimport move
 
-from .structs cimport MorphAnalysisC
 from .strings cimport StringStore
 from .typedefs cimport attr_t, hash_t
 
 
+cdef cppclass Feature:
+    hash_t field
+    hash_t value
+
+    __init__():
+        this.field = 0
+        this.value = 0
+
+
+cdef cppclass MorphAnalysisC:
+    hash_t key  
+    vector[Feature] features
+
+    __init__():
+        this.key = 0
+
 cdef class Morphology:
-    cdef readonly Pool mem
     cdef readonly StringStore strings
-    cdef PreshMap tags # Keyed by hash, value is pointer to tag
-
-    cdef MorphAnalysisC create_morph_tag(self, field_feature_pairs) except *
-    cdef int insert(self, MorphAnalysisC tag) except -1
+    cdef unordered_map[hash_t, unique_ptr[MorphAnalysisC]] tags
```

Comment:
Just wanted to avoid the overhead of atomic increment/decrement ops required to operate the reference counter in `std::shared_ptr`. I considered avoiding the pointer indirection entirely, but Cython's limited support for move semantics meant that we can't construct the `MorphAnalysisC` object in-place in the `map` and will have to suffer allocations, deallocations multiple temporaries and at least one copy.

I do agree with your reasoning, though. I can do some more profiling to determine if the `shared_ptr` overhead is small enough to not matter.